tests/test_utils.py

```diff
@@ -65,31 +69,88 @@
+                Tensor(PauliZ(wires=[0]), PauliX(wires=[1])),
+                Tensor(PauliZ(wires=[0]), PauliY(wires=[1])),
+            ],
+        ),
+    ],
+]
 
 
 class TestDecomposition:
     """Tests the decompose_hamiltonian function"""
 
-    @pytest.mark.parametrize("H", test_Hamiltonians)
-    def test_decomp(self, H):
-        """Tests that decompose_hamiltonian successfully decomposes Hamiltonians into Pauli matrices"""
-        N = int(np.log2(len(H)))
-        coeff, combs = pu.decompose_hamiltonian(H)
-        result = np.zeros((2 ** N, 2 ** N))
-        matrices = []
-        for term in combs:
-            base = [qml.Identity] * N
-            if term.num_wires == N:
-                matrices.append(term.matrix)
-            else:
-                base[term.wires[0]] = term.__class__
-                matrices.append(
-                    functools.reduce(operator.matmul, [t(i) for i, t in enumerate(base)]).matrix
+    def test_wrong_dimension(self):
+        with pytest.raises(
+            ValueError,
+            match=re.escape("Hamiltonian should have shape (2^n, 2^n), for any qubit number n>=1"),
+        ):
+            pu.decompose_hamiltonian(np.ones((3, 3)))
+
+    def test_not_hermitian(self):
+        with pytest.raises(ValueError, match="The Hamiltonian is not Hermitian"):
+            pu.decompose_hamiltonian(np.array([[1, 2], [3, 4]]))
+
+    @pytest.mark.parametrize("H", test_hamiltonians)
+    def test_decomposition(self, H):
+        """Tests that decompose_hamiltonian successfully decomposes Hamiltonians into a
+        linear combination of Pauli matrices"""
+        test, expected = H
+        decomposed = pu.decompose_hamiltonian(test)
+
+        decomposed_coeff, decomposed_obs = decomposed
+        expected_coeff, expected_obs = expected
+
+        assert len(decomposed_obs) == len(expected_obs)
+        assert len(decomposed_coeff) == len(expected_coeff)
+
+        assert np.allclose(decomposed_coeff, expected_coeff)
+
+        n_terms = len(decomposed_obs)
+
+        linear_comb = np.zeros(test.shape, dtype=np.complex128)
+        for term in range(n_terms):
+            assert type(decomposed_obs[term]) == type(expected_obs[term])
+
+            if isinstance(decomposed_obs[term], Tensor):
+                obs = zip(decomposed_obs[term].obs, expected_obs[term].obs)
+                assert all(np.allclose(o[0].matrix, o[1].matrix) for o in obs)
+
+                linear_comb += decomposed_coeff[term] * functools.reduce(
+                    np.kron, [o.matrix for o in decomposed_obs[term].obs]
                 )
-        for term in zip(coeff, matrices):
-            result = result + term[0] * term[1]
-        assert (H == result).all
+            else:
+                assert np.allclose(decomposed_obs[term].matrix, expected_obs[term].matrix)
+                linear_comb += decomposed_coeff[term] * decomposed_obs[term].matrix
+
+        assert np.allclose(test, linear_comb)
```

Comment:
Very thorough testing :) I agree with @trbromley's comment about splitting up this particular tests into separate tests for readability and maintainability

Reply:
Sounds good, I'll refactor as you both suggested.